common code refactoring

colinsurprenant opened the pull request:
This is work toward Data Streams integration elastic/logstash#12178

Specific elasticsearch methods are now into the main elasticsearch class file and shared methods have been moved into the PluginMixins::ElasticSearch::Common namespace. The license checking code has been externalized and can now be specified as argument to the `build_client` method.
    
This is code refactorting that has no end-user impact.

yaauie commented on lib/logstash/outputs/elasticsearch/http_client/pool.rb:
```diff
@@ -282,19 +284,10 @@ def healthcheck!
-            end
+
+            # defined in license_check.rb
+            license_check!(url, meta)
```
Maybe a pre-flight proc (`connection_preflight`?) could be registered by the thing creating the client pool, which would enable different common consumers to do their own pre-flights, perhaps including their own specific license-checking behaviour or registering a no-op preflight if there is no extra steps needed.

colinsurprenant replied:
🤔

colinsurprenant replied:
The license checking logic is badly intertwined in the Pool logic and I am a bit worried about performing this refactor at this stage. I agree that it would be a positive code cleanup but I am unsure it is worth the investment now-now just for the benefit of making the data_streams output work. This proposed solution is rather non-intrusive and I will add more comments about the intent. Maybe keep a tab on this as a future refactoring followup?

colinsurprenant replied:
you can disregard this above comment, I moved that discussion at the top level.

yaauie reviewed (commented):
Most things seem to fall on the "right" side of the re-grouping, at least in terms of making a common reusable core goes.

I'm not really sure about where the License Checker should "live", but it seems both deeply tied to both the pool's implementation and the Elasticsearch output's requirements. Since we know that the requirements of other consumers using this common core will have differing requirements, maybe we should build out a preflight abstraction so that a consumer can define its own preflight check.

I mashed my keyboard a bit to prototype the kind of thing I'm envisioning. The setup is a bit verbose, but the usage ends up being pretty readable https://gist.github.com/yaauie/d73c4ab4284909cd3e0a5309ea3cd938

colinsurprenant commented on lib/logstash/outputs/elasticsearch/http_client/pool.rb:
```diff
@@ -69,7 +71,7 @@ def initialize(logger, adapter, initial_urls=[], options={})
     end
 
     def oss?
```
+1

colinsurprenant commented on lib/logstash/outputs/elasticsearch/license_check.rb:
```diff
@@ -0,0 +1,29 @@
+        # @param url [LogStash::Util::SafeURI]
+        # @param meta [Hash]
+        def license_check!(url, meta)
+          if oss? || valid_es_license?(url)
```
yeah,  I considered this, and settled for only exposing just a *main* `license_check!` method because I do not think more will be required on the data_streams side to implement the check but I think `oss?` and `valid_es_license?` should probably go there too. I also wanted to avoid performing too deep a refactoring just for a slight change in the check logic. This struck me as an acceptable way to externalize the check - I will definitely add a *big* comment about its usage and context; the fact it is only meant to live within the Pool class. 
I know this a rather loose integration and I'll see how I can add relevant specs to make sure the *api* for this mixin is somehow protected.

colinsurprenant replied:
So I opted to also extract `oss?`, `valid_es_license?`, `log_license_deprecation_warn` together with the main entry point ` license_check! `.

colinsurprenant commented:
@yaauie For some reason I did not catch your top level comment above and only followed up the inline comments.

I agree in essence with your license checking proposal. As you know the license checking logic is badly intertwined in the Pool logic and I am a bit worried about performing this refactor at this stage with the added risks. I agree that it would be a positive code cleanup but I am unsure it is worth the investment now-now just for the benefit of making the data_streams output work. 

This proposed solution is rather non-intrusive with minimal code change. I suggest we followup with your license checking proposal after the first data_streams iteration?

colinsurprenant commented:
WDYT about renaming  `Common` module `common.rb` into `ApiMethods`  `api_methods.rb` similar to what we did for `ApiConfigs` ?

colinsurprenant commented:
This is ready be moved from Draft to Ready for review.
I believe the pending issues are:
- resolution for the license checking conversation
- suggestion on renaming the `Common` module into `ApiMethods`

yaauie reviewed (commented):
If the Elasticsearch Output's license checker is statically embedded in the pool implementation (e.g., by including a module into the pool's class), we cannot reuse the pool from another plugin that has its own license-checking logic. This is the primary reason why I think we need to pass a pre-flight checker _to_ the pool via the client builder, so that each consumer of the pool can define its own license checking independently of the pool's implementation.

colinsurprenant commented:
@yaauie as discussed, thanks for the heads up on this, I now see the potential problem with my suggestion and definitely need to refactor it.

colinsurprenant commented:
@yaauie I refactored the license checking to use a *proper* class and not a mixin which will solve the problem of collision when/if both the elasticsearch and the data_streams outputs are used in the same VM. This license checking  class is required in the top `elasticsearch.rb`  and passed from the `build_client` call down to `Pool#initialize` where the license check class is instantiated if provided otherwise the local `license_check.rb` file is required and that class is used for BWC for other users of the `build_client` methods that assume to be using the elasticsearch output plugin license checking behaviour. 

The license checking class, similar to the prior mixin is made to minimized change as much as possible. Other than minimizing risk, one of the reason is that all this code between the pool, http_client, http_client_builder and the related specs is really entangled and a refactor similar to what you suggested would require quite a deeper refactoring and lots of specs changes and at this point I am not sure we should take that on plus the related risks.

yaauie commented on lib/logstash/outputs/elasticsearch/http_client/pool.rb:
```diff
@@ -282,19 +282,10 @@ def healthcheck!
-              log_license_deprecation_warn(url)
-              meta[:state] = :alive
-            end
+
+            # license_check in mixed in from LogStash::Outputs::ElasticSearchPoolMixin::LicenseChecker
+            # separately defined in the elasticsearch output and the elasticsearch_data_streams output.
+            @license_checker.license_check!(url, meta)
```
if we only had a pre-flight checker as I had suggested, that was given a reference to this pool and the `url` and simply returned true or false but _was not expected to manipulate pool internals_, then we could worry about pool internals here, in the pool implementation, and not in different consumers of this API in different code-bases.

```suggestion

            if @preflight.check?(this, url)
              meta[:state] = :alive
            else
              logger.warn("Connection failed pre-flight check", :url => url)
            end
```

The Elasticsearch Output then, could define its own pre-check that considered all of the stages in its license check without having to care about pool internals:

https://gist.github.com/yaauie/d73c4ab4284909cd3e0a5309ea3cd938#file-preflight_check-rb-L89

yaauie commented on lib/logstash/outputs/elasticsearch/http_client/pool.rb:
```diff
@@ -66,12 +67,19 @@ def initialize(logger, adapter, initial_urls=[], options={})
+        # normally this class is passed from the Commons#build_client method to allow
+        # alternate license checking logic form other implementations like the data streams output.
+        require "logstash/outputs/elasticsearch/license_check"
+        @license_checker = LogStash::ElasticSearchOutputLicenseChecker.new(self, logger)
```
This may break Logstash core's central pipeline management, which already reaches into the Elasticsearch Output to `build_client` without any args. I think we need to have a no-op _always-success_ implementation if no checker is given.

yaauie reviewed (commented):
A big part of the motivation to my previous recommendation (pre-flight), was to separate concerns. Requiring the license check's implementation to manipulate the pool's internals is dangerous, because it effectively makes the internal details of the pool implementation part of its API.

If the pre-flight is _only_ responsible for emitting `true`/`false` and its `check?` method has access to the pool and base url being pre-flighted with _no expectation that it be used to mutate said pool or url_, then any implementation can tell us whether it is licensed _without_ having to care about pool internals. This is especially important because the first consumer of these APIs we plan creating is going to live in a separate repository, with separate versioning and release timing, and very little ability to positively ensure lock-step matching of versions when installed into Logstash.

I think that a pre-flight as I suggested can be plumbed through in very much the same way you have plumbed a license checker through, except that when no pre-flight is given we should rely on a noop implementation whose `Preflight#check?` always returns `true` (the `LogStash::Outputs::Elasticsearch#build_client` is used from logstash core's central pipeline management, so we want to avoid changes that would affect a no-arguments version of that method).

colinsurprenant commented:
@yaauie I agree with your assessment, my goal was do really minimize change. I also agree we can probably improve without too much code disruption. 

WRT to core's pipeline management which relies on the current `build_client` implementation, it is my understanding that it also relies on the same license checking logic as the current elasticsearch output and thus would not require a no-op default upon a no-arg usage on `build_client` no?

yaauie commented:
> WRT to core's pipeline management which relies on the current `build_client` implementation, it is my understanding that it also relies on the same license checking logic as the current elasticsearch output and thus would not require a no-op default upon a no-arg usage on `build_client` no?

Central Management does its own license checking _once the pool is established_, and the flow includes helpful hints to a user when they connect to an Elasticsearch that is unlicensed. Right now, the Elasticsearch Output does _warn_ when connecting a default-distribution Logstash to an unlicensed Elasticsearch (and Central Management gets this warning "for free" whether it wants it or not), but in 8.0 we intend to change that behaviour to _block_ the connection from being usable. When we do this, if Logstash's Central management continues to use the Elasticsearch Output's `build_client` to get its connection, its license checker will be unable to connect and our helpful path will be effectively unreachable.

I believe that the current license check warning _missed_ the fact that the pool was also being used by Logstash core. It is really hard to keep track of what things are reaching into this plugin and depend on the specific shape of its internals, and because we cannot lock-step this plugin with the things that reach into it, it is very important that we establish clear boundaries. Refactoring this license check _up_ to the specific consumer that needs it (and back-tracking its accidental exposure out of the consumers that don't) is the only way I see to make this maintainable, especially as we look to add consumers.

colinsurprenant commented:
@yaauie 

**1-** IMO refactoring core's usage of `build_client` is out of scope and my intention is to keep the no-arg `build_client` compatible so that core does not need to be updated with the new version of the plugin and there is very minimal behaviour change.

There are 2 places where `build_client` is used in core:

- in x-pack/lib/license_checker/license_reader.rb: https://github.com/elastic/logstash/blob/eed7a1538ce71069da12a48f9bb7bf37673364d0/x-pack/lib/license_checker/license_reader.rb#L54-L64

- in x-pack/lib/config_management/elasticsearch_source.rb: https://github.com/elastic/logstash/blob/eed7a1538ce71069da12a48f9bb7bf37673364d0/x-pack/lib/config_management/elasticsearch_source.rb#L115-L126

Both places display this comment:

>      # This is a bit of a hack until we refactor the ElasticSearch plugins
>      # and extract correctly the http client, right now I am using the plugins
>      # to deal with the certificates and the other SSL options
>      #
>      # But we have to silence the logger from the plugin, to make sure the
>      # log originate from the `ElasticsearchSource`

The good news here is that once we are done with this PR, we will be able to update core x-pack code in both these places to use the extracted common code here and properly build a client with an appropriate license checking class without having to instantiate the whole plugin. 

2- I pushed changes to remove dependencies to the Pool class form the license checker. I believe we are getting closer to what you would like to see but also by keeping very close & similar implementation and license checking flow. The only functional difference now is that the HTTP request to fetch the license is done systematically on every `healthcheck!` call while previously  it was only done is logstash was not OSS. 

Again I understand this is not exactly how you see how the license checker should be ideally structured but this implementation has the benefit of not disturbing the specs too much which is preferable because if you refactor the code and also have to significantly change the specs then you increase the risk of introducing behaviour changes.

yaauie reviewed (commented):
> IMO refactoring core's usage of build_client is out of scope and my intention is to keep the no-arg build_client compatible
> [...]
> The good news here is that once we are done with this PR, we will be able to update core x-pack code in both these places to use the extracted common code here and properly build a client with an appropriate license checking class without having to instantiate the whole plugin.

My point is that we _must not_ do connection-level license checking in the Logstash core usages of this plugin's shared pool, because those usages require a valid connection to walk people toward the correct configuration. We _accidentally_ do so now, which is a bug (thankfully we only emit a warning and haven't yet made the breaking change to prevent the connection from being added to the pool). We have the opportunity to not perpetuate this bug.

> This implementation has the benefit of not disturbing the specs too much which is preferable because if you refactor the code and also have to significantly change the specs then you increase the risk of introducing behaviour changes.

Exposing an API is an inflection point, and failure to get that API right injects a lot of risk that we cannot ignore.

I know we're doing a lot of back-and-forth here, and it is probably because I failed to fully-explain _why_ I was proposing the pre-flight interface in the shape I proposed it.

 - the `Preflight#check?` method in my proposal was carefully crafted to take this `Pool` and the `url` as arguments. This allows any implementation to do whatever pre-flight checking it needs to do, including but not limited to license checks. When a consumer of this API can only be used with specific versions of Elasticsearch (e.g., because it depends on a feature introduced mid-major) or with specific Elasticsearch features enabled (e.g., data streams) it can codify such requirement in its pre-check. I believe that this is a boundary that solves our current use-case (the Elasticsearch Output), while providing sufficient room for planned and unplanned future work.

I've put together a PR against your branch to show what I'm talking about. It includes the existing on-pool spec validation of the ElasticsearchOutputLicenseChecker, adds unit-specs for the pool using a license checker that _doesn't_ care bout the license checker's internals, and includes a noop license checker implementation to be used by external consumers like Logstash core when `build_client` is called without specifying a license checker.

https://github.com/colinsurprenant/logstash-output-elasticsearch/pull/1

colinsurprenant commented:
@yaauie 

> My point is that we must not do connection-level license checking in the Logstash core usages of this plugin's shared pool, because those usages require a valid connection to walk people toward the correct configuration. We accidentally do so now, which is a bug (thankfully we only emit a warning and haven't yet made the breaking change to prevent the connection from being added to the pool). We have the opportunity to not perpetuate this bug.

Ok sure, but I don't think changing this behaviour in v7 is a good idea. For v8  core x-pack can pass in a no-op license checking class to its `build_client` and we can add a simple check that is a no-op class is passed, the whole license checking section will be skipped in `Pool`. Not a problem.

colinsurprenant commented:
@yaauie about your proposal https://github.com/colinsurprenant/logstash-output-elasticsearch/pull/1 is essentially what I had right before I removed passing `Pool` the the license checker constructor to avoid relying on the `Pool` internals. I am on the fence about that but I think passing in `Pool` to the license checker is in fact cleaner and will avoid dealing with possible unwated `get_license` call from the `heath_check!` method. 

As for your proposal for using the `NoopLicenseChecker` when using the no-arg `build_client` I think we should not do it in v7 as I mentioned above. Instead I think we should pass the `NoopLicenseChecker`  to the x-pack  `build_client` calls in a future core PR for either v8 or for the next minor. But changing no-arg `build_client` behaviour in v7 is risky IMO. 

+1 on the changes you suggest, minus the change in behaviour of the no-arg `build_client`.

yaauie commented:
> I don't think changing this behaviour in v7 is a good idea

Connection-level license checking was implemented on the Elasticsearch Output to deprecate a specific use of the Elasticsearch Output (https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/875), and was _accidentally_ consumed by the Logstash core classes that reach in and use the Elasticsearch Output's `build_client`. The bug causes Logstash core to make an additional API call per connection to get the license, whose one and only effect is to log a message that is not relevant or accurate to those use-cases. I do not see the value in intentionally perpetuating this bug.

colinsurprenant commented:
@yaauie I understand but still I don't think we should change that behaviour which will apply to any logstash version installing that new version of the elasticseach output plugin, I think this is risky and that we should only introduce the explicit no-op license checker into `build_client` moving forward in future versions of logstash.

colinsurprenant commented:
@yaauie I pushed the changes as suggested and discussed above.

colinsurprenant commented:
@yaauie following up on our discussion and your pointers on the x-pack license checking logic, AFAICS both the monitoring and management are in fact doing the license checking outside the connection creation of `build_client`. 

The x-pack license checking in https://github.com/elastic/logstash/tree/master/x-pack/lib/license_checker  is used in both places :

- Config Management:  `elasticsearch_source.rb` launches the license checker in the constructor https://github.com/elastic/logstash/blob/ca81a8f4a32457a7c100a4712c7c5a4c9b5c2faa/x-pack/lib/config_management/elasticsearch_source.rb#L37-L41. 

- Monitoring: `internal_pipeline_source.rb` does the same https://github.com/elastic/logstash/blob/ca81a8f4a32457a7c100a4712c7c5a4c9b5c2faa/x-pack/lib/monitoring/internal_pipeline_source.rb#L22

So in all x-pack cases `build_client` is only used for connection-level to ES and all license checking logic is done outside using the `LicenseChecker` module.

So I am +1 on making the no-op license checker the default upon a no-arg `build_client` call which will provide a cleaner experience for existing installed LS upgrading to the newer elasticsearch output.

yaauie reviewed (commented):
:tada: this is looking like all the necessary separations and should set us up for success moving forward.

I'm not particularly comfortable about dropping things directly into the `LogStash` namespace, and would prefer if file paths and class names generally matched.

We have prior art for plugins that publish sharable code doing so in the `LogStash::PluginMixins` namespace. Perhaps with that in mind:

 1. The Elasticsearch Output's license checker could live as `LogStash::Outputs::Elasticsearch::LicenseChecker` in `lib/logstash/outputs/elasticsearch/license_checker.rb`, which would clearly identify it as being part of the ES Output plugin.
 2. The shared noop license checker could live as `LogStash::PluginMixins::Elasticsearch::LicenseChecker::NOOP_INSTANCE` or similar in `lib/logstash/plugin_mixins/elasticsearch/license_checker.rb`
 3. The shared common module you extracted could live as `LogStash::PluginMixins::Elasticsearch::Common` in `lib/logstash/plugin_mixins/elasticsearch/common.rb`

While the pool and http client various other code is also now effectively shared by nature of the common module exposing them, there certainly is risk in change file paths, and that probably outweighs the benefit of a wholesale rearrangement.

colinsurprenant commented:
@yaauie +1 on the namespacing.

WDYT about #973 `ApiConfig` ? We should probably also put that into this new mixin path? The current namespace is  `module LogStash; module Outputs; class ElasticSearch; module APIConfigs`  which is weird, that does not need to be into the `ElasticSearch` class, we could have it into a all-module namespace into the mixin path.

colinsurprenant commented:
@yaauie I pushed the namespace shuffle.

jsvd commented on lib/logstash/plugin_mixins/elasticsearch/common.rb:
```diff
@@ -15,51 +18,6 @@ module Common
     # will never succeed.
     VERSION_TYPES_PERMITTING_CONFLICT = ["external", "external_gt", "external_gte"]
```
I don't think this needs to be in common anymore: it will be used in existing ES output, but not in data streams.

It feels like the [event_action_tuple](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/efc1c069d04444f77db1549d95e4f06131497f11/lib/logstash/plugin_mixins/elasticsearch/common.rb#L33) method doesn't really belong in common as most of it is around doing stuff that data streams wont do: parent/child, update, version,etc.

colinsurprenant replied:
ok let me double-check for `event_action_tuple `

colinsurprenant replied:
I moved the whole `event_action_tuple` + friends out of commons.

colinsurprenant replied:
searching for `VERSION_TYPES_PERMITTING_CONFLICT` in all plugins or logstash repo I can't fin any reference/usage of that. Will remove.

colinsurprenant commented on lib/logstash/outputs/elasticsearch/http_client/pool.rb:
```diff
@@ -282,19 +283,9 @@ def healthcheck!
-            end
+
+            alive = @license_checker.appropriate_license?(self, url)
+            meta[:state] = alive ? :alive : :unlicensed
```
so would `:dead` be the right state for this then?  
The valid states and their meanings should ideally be encoded somewhere as constants or such. No sure we want to introduce that refactor here but we should clean that up at some point.

colinsurprenant replied:
So looking more closely at the code the 2 places where meta is really used is:

- https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/efc1c069d04444f77db1549d95e4f06131497f11/lib/logstash/outputs/elasticsearch/http_client/pool.rb#L267 

`healthcheck!` cycles through all `@url_info` with a meta other than `:alive` to see if it can be reused, so here any value, like `:unlicensed` is ok, a health check will be performed periodically on these.

- https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/efc1c069d04444f77db1549d95e4f06131497f11/lib/logstash/outputs/elasticsearch/http_client/pool.rb#L441-L451

`get_connection` will pick a connection if it is not `:dead`. Here I am not sure if the logic should not be to only consider the `:alive` ones instead of the not `:dead` ones if we introduce another state like `:unlicensed`. The question is: would is make sense for `get_connection` to return a `:unlicensed` connection? I wouldn't think so. 

Currently a `:dead` connection is set upon `HostUnreachableError` in https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/efc1c069d04444f77db1549d95e4f06131497f11/lib/logstash/outputs/elasticsearch/http_client/pool.rb#L402-L405.

So with this, it seems that we could also simply set the `meta[:state] = :dead` instead of a new `:unlicensed` state and we would not have to change anything else in the logic. The above 2 places where meta is looked at would work; `healthcheck!`  would perform a check on a `:dead` url and `get_connection`  would not pick it up if it was maked as `:dead`. 

LMKWYT.

colinsurprenant commented on lib/logstash/plugin_mixins/elasticsearch/common.rb:
```diff
@@ -77,7 +35,7 @@ def event_action_tuple(event)
       params = {
         :_id => @document_id ? event.sprintf(@document_id) : nil,
-        :_index => event.sprintf(@index),
+        :_index => build_event_index(event),
```
`build_event_index` is not defined in commons so it must be defined in the actual plugin and the implementation is different in both cases. But maybe as you suggested above, the whole `event_action_tuple` could actually be removed from commons instead. it's data stream implementation will be much simpler actually.

colinsurprenant replied:
I moved the whole `event_action_tuple` + friends out of commons, dud not make sense to have it there.

colinsurprenant commented on lib/logstash/plugin_mixins/elasticsearch/common.rb:
```diff
@@ -121,6 +79,29 @@ def event_action_tuple(event)
+    # Note that this methods may sets the @user, @password, @hosts and @client ivars as a side effect.
+    # @param license_checker [#appropriate_license?] An optional license checker that will be used by the Pool class.
+    # @return [HttpClient] the new http client
+    def build_client(license_checker = nil)
```
good point. but `metric` is a base plugin getter, IMO `build_client` should not have been calling such a plugin getter  in the first place; passing a metric as parameter would have been cleaner. I am not sure we we should continue down that path?

this process here is a very first step at *trying* to untangle more *generic* code from the more specific plugin code. This metric thing is a perfect example of not properly separating concerns (this is totally blameless, this plugin has a huge history). `build_client` is being used in other contexts than an actual plugin and if this code had been better encapsulated it would have been a lot easier.

colinsurprenant commented:
@jsvd I think I addressed all your review comments. Only thing I'd ask you to double check is the `meta[:state] = :dead` instead of `:unlicensed`. IMO this is better. 

LMKWYT.

jsvd commented:
Looking at the test failures, can you check if the OSS test that took 6h was a hiccup somehow? Looking at the output log from travis it all seemed to work well?

colinsurprenant commented on lib/logstash/outputs/elasticsearch.rb:
```diff
@@ -322,4 +370,116 @@ def self.oss?
+
+  private
+
+  def discover_cluster_uuid
+    return unless defined?(plugin_metadata)
+    cluster_info = client.get('/')
+    plugin_metadata.set(:cluster_uuid, cluster_info['cluster_uuid'])
+  rescue => e
+    # TODO introducing this logging message breaks many tests that need refactoring
+    # @logger.error("Unable to retrieve elasticsearch cluster uuid", error => e.message)
+  end
+
+  def successful_connection?
+    !!maximum_seen_major_version
+  end
```
hmm good question; currently `discover_cluster_uuid` is called in the context of the template installer 

https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/9451cfbb02297c736ee47c59da1204650da68f22/lib/logstash/outputs/elasticsearch.rb#L436-L448

and I assumed this was rather specific to the ES output only - this `setup_after_successful_connection` method is only called from ES output `register` and not from the DS output.

Is this something the DS output will require? the the `:cluster_uuid` in the plugin metadata useful in other places? https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/9451cfbb02297c736ee47c59da1204650da68f22/lib/logstash/outputs/elasticsearch.rb#L379

jsvd replied:
cluster_uuid has been used for linking the logstash instance with the ES cluster for monitoring purposes when metric collection is done externally. So we still need to provide this metadata when the plugin uses datastreams

colinsurprenant commented:
@jsvd yeah I looked and did not understand a) why it took 6hrs and b) why is shows a failed state although all tests passed and the exit code was 0 🤔  
I need to  push another commit, we'll see how it goes.

colinsurprenant commented:
@jsvd the previous build error is gone but now there's a new one which seems unrelated. 

```
Bundler could not find compatible versions for gem "sinatra":
  In Gemfile:
    logstash-core java was resolved to 8.0.0, which depends on
      sinatra (~> 2) java
Could not find gem 'sinatra (~> 2)', which is required by gem 'logstash-core',
in any of the sources.
ERROR: Service 'logstash' failed to build: The command '/bin/sh -c bundle install --with test ci' returned a non-zero code: 6
The command ".ci/docker-setup.sh" failed and exited with 1 during .
```

I am trying to reproduce locally.

colinsurprenant commented:
Can't reproduce locally, I am restarting the job to see.

colinsurprenant commented:
Green ✅

jsvd reviewed (approved):
I ran a few tests with Elasticsearch OSS and default distro and it all seems to work fine.

Overall I'm not too giddy about the license checker injection, but we first need to stop using plugin internals for Central Management.

jsvd commented:
@colinsurprenant I missed the uuid comment, which I think justifies a last change to have the successful_connection and discover_cluster_uuid apparatus in the common file https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/976#discussion_r532655961

colinsurprenant commented:
@jsvd ok, makes sense for UUID - good catch.

colinsurprenant commented:
@jsvd I pushed the changes related to `discover_cluster_uuid` as discussed. This could probably benefit some refactoring in the future but I think this change will make that possible and minimized changes.

jsvd reviewed (approved):
LGTM after green build

colinsurprenant commented:
Thanks @yaauie @jsvd for the reviews. This is merged and will be part of 10.8.0.
